### PR TITLE
Add explicit error message when specifying default_return_value on void function call

### PR DESCRIPTION
**Pull request by benber86**
### What I did

When specifying the `default_return_value` on a call to a void external function, the compiler currently only returns the following error message `AttributeError: 'NoneType' object has no attribute 'compare_type'`. 
I added a more explicit exception so that the error can be caught and addressed more easily.

### How I did it

I added a check for a return type when `default_return_value` is set. If there's no return type, an `ArgumentException` is raised.

### How to verify it

Compiling this contract:

```
interface FooBar:
    def foo(): payable

@external
def bar():
    FooBar(empty(address)).foo(default_return_value=True)
```

will now generate the following error:

```
vyper.exceptions.ArgumentException: default_return_value specified but foo is void
  contract "dist/simple.vy:6", function "bar", line 6:52 
       5 def bar():
  ---> 6     FooBar(empty(address)).foo(default_return_value=True)
  -----------------------------------------------------------^
       7
```

### Commit message

Exception when setting default_return_value on void function call

### Description for the changelog

Add explicit error message when specifying default_return_value on void function call

### Cute Animal Picture

image

### CODE REVIEW

**Review by charles-cooper — COMMENTED** [review]
is there any case where `require_return=` should not be set to true?

File: vyper/semantics/types/function.py [hunk 1]
```diff
@@ -478,6 +478,10 @@ def fetch_call_return(self, node: vy_ast.Call) -> Optional[BaseTypeDefinition]:
         for kwarg in node.keywords:
             if kwarg.arg in self.call_site_kwargs:
                 kwarg_settings = self.call_site_kwargs[kwarg.arg]
+                if kwarg_settings.require_return and self.return_type is None:
```
Comment (charles-cooper) [inline vyper/semantics/types/function.py:481]: probably this branch is cleaner if we don't modify the KwargSettings data structure.
```suggestion
                if kwarg.arg == "default_return_value" and self.return_type is None:
```

**Comment by benber86** [comment]
> is there any case where `require_return=` should not be set to true?

Not for the `default_return_value` kwarg specifically. Would you prefer something like `if kwarg.arg == "default_return_value"` over having a flag for it?

**Review by fubuloubu — APPROVED** [review]

File: vyper/semantics/types/function.py [hunk 2]
```diff
@@ -478,6 +478,10 @@ def fetch_call_return(self, node: vy_ast.Call) -> Optional[BaseTypeDefinition]:
         for kwarg in node.keywords:
             if kwarg.arg in self.call_site_kwargs:
                 kwarg_settings = self.call_site_kwargs[kwarg.arg]
+                if kwarg.arg == "default_return_value" and self.return_type is None:
+                    raise ArgumentException(
+                        f"{kwarg.arg} specified but {self.name} is void", kwarg.value
```
Comment (charles-cooper) [inline vyper/semantics/types/function.py:483]:
```suggestion
                        f"{kwarg.arg} specified but {self.name} does not return anything", kwarg.value
```

**Review by fubuloubu — APPROVED** [review]

**Review by charles-cooper — APPROVED** [review]

File: vyper/semantics/types/function.py [hunk 3]
```diff
@@ -478,6 +478,10 @@ def fetch_call_return(self, node: vy_ast.Call) -> Optional[BaseTypeDefinition]:
         for kwarg in node.keywords:
             if kwarg.arg in self.call_site_kwargs:
                 kwarg_settings = self.call_site_kwargs[kwarg.arg]
+                if kwarg.arg == "default_return_value" and self.return_type is None:
+                    raise ArgumentException(
+                        f"{kwarg.arg} specified but {self.name} does not return anything", kwarg.value
```
Comment (charles-cooper) [inline vyper/semantics/types/function.py:483]: slightly clearer error message:
```suggestion
                        f"{kwarg.arg}= specified but {self.name} does not return anything", kwarg.value
```

**Review by charles-cooper — APPROVED** [review]

**Pull request merged**